### DISCUSSION TITLE
How to apply a ruleset only on src/main/*.java files that do not match *Foo*.java?

### DISCUSSION THREAD

**Discussion by cinlloc** [pmd/pmd#5217] [Q&A] [answered]
Hi, I tried to write a ruleset that match only java files that are in `src/main` and does not match *Foo*. Seems to be impossible with include/exclude PMD mechanism (exclusions are applied if file does not match a inclusion file). What I tried to use something like:

```xml
...
<exclude-pattern>.*</exclude-pattern>
<include-pattern>.*/src/main/java/com/laroueverte/.*(?!Foo).*.java</include-pattern>
...
```

But the `(?!Foo)` exclusion regex does not seems to work.

**Comment by jsotuyod** [comment]
I've just tested it and it works as expected… how are you running PMD?

**Reply by cinlloc** [reply]
Here's the Java project showing the issue : https://github.com/cinlloc/pmd-pattern-issue-demo

I run PMD two ways: with Maven plugin `mvn compile pmd:check` and with IntelliJ plugin, both give the same result.  

I expect only one validation warning on `Bar.java`, but I have two (`Foo.java` and `Bar.java`, i.e. `Foo.java` is not excluded as I expect).

The regex seems to be syntactically correct for the parser (no error at execution), but the `(?!Foo)` part seems to be ignored.

**Comment by adangel** [comment, marked as answer]
> But the (?!Foo) exclusion regex does not seems to work.

Yes, that seems to be the case. But it has something to do with regex, rather than PMD:
https://regex101.com/r/F4JBok/1

An easier way to achieve what you want: Just use exclusion without inclusion:

```
<exclude-pattern>.*Foo.java|.*/src/test/java/.*</exclude-pattern>
```

You could remove the "src/test/java"-part, if you set the maven-pmd-plugin parameter `<includeTests>false</includeTests>`.

**Reply by jsotuyod** [reply]
Alternatively, a proper include regexp would be:

```
.*/src/main/java/.*/(?![^/]*Foo[^/]*)[^/]*\.java
```

Which will not include any file with `Foo` in the name.
The main issue is the wildcard matches around the negative lookahead in the original regex are basically annulling it, as the regex can decide any number of chars to be matched on each end, so that Foo is never found in the middle…

For instance, given `Foo.java`, if the first `.*` matches 0 chars, Foo is found, and the match is failed, but the Regex can be retried by matching 1 char instead, so the first `.*` matches `F`,  the negative lookahead doesn't match `Foo`,  and the final `.*` can match `oo`, so it's a match.

**Reply by cinlloc** [reply]
Ok. Thanks for the answer!
